File: files/problems/cpp/generate-parentheses.cpp
// by @codeAbinash

class Solution {
   public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string str = "";
        generate(ans, str, n, n);
        return ans;
    }
    void generate(vector<string>& ans, string& str, int open, int close) {
        if (open == 0 && close == 0) {
            ans.push_back(str);
            return;
        }
        if (open > 0) {
            str.push_back('(');
            generate(ans, str, open - 1, close);
            str.pop_back();
        }
        if (close > 0 && open < close) {
            str.push_back(')');
            generate(ans, str, open, close - 1);
            str.pop_back();
        }
    }
};

File: leetcode/problems/rs/generate-parentheses.rs
// by @codeAbinash
// Time : O(n)
// Space : O(n)

pub struct Solution;

impl Solution {
   pub fn generate_parenthesis(n: i32) -> Vec<String> {
      let mut res = vec![];
      let mut stack = vec![];
      Self::backtrack(&mut res, &mut stack, n, 0, 0);
      res
   }

   fn backtrack(res: &mut Vec<String>, stack: &mut Vec<char>, n: i32, open: i32, close: i32) {
      if stack.len() == n as usize * 2 {
         res.push(stack.iter().collect());
         return;
      }

      if open < n {
         stack.push('(');
         Self::backtrack(res, stack, n, open + 1, close);
         stack.pop();
      }

      if close < open {
         stack.push(')');
         Self::backtrack(res, stack, n, open, close + 1);
         stack.pop();
      }
   }
}
